File: binary_search/binary_search.py
"""
Este es un módulo de búsqueda binaria.
Proporciona la función 'binary_search' para realizar una búsqueda binaria en una lista ordenada.
"""
def binary_search(lista, item):
    """
     Realiza una búsqueda binaria en una lista ordenada para encontrar el índice del elemento dado.

    Args:
        list (list): La lista **ordenada** en la que se realizará la búsqueda.
        item: El elemento que se desea encontrar.

    Returns:
        int or None: El índice del elemento en la lista si se encuentra, None si no se encuentra.
    """
    low = 0
    high = len(lista) - 1
    pasos = 0

    while low <= high:
        mid = (low + high) // 2
        guess = lista[mid]
        pasos +=1

        if guess == item:
            print(f'El total de pasos fue: {pasos}')
            return mid
        if guess > item:
            high = mid - 1
        else:
            low = mid + 1
    return None


my_list = [1, 3, 5, 7, 9]

print(binary_search(my_list, 3))

File: binary_search/binary_search.js
/**
 * Realiza una búsqueda binaria en un array ordenado para encontrar el valor objetivo.
 * @param {array} array                       - El array ordenado en el que se realizará la búsqueda.
 * @param {*} targetValue                     - El valor objetivo que se desea encontrar.
 * @returns {number}                          - El índice del valor objetivo en el array, o -1 si no se encuentra.
 */
var doSearch = function(array, targetValue){
    var min = 0;
    var max = array.length - 1;
    var guess;
    var totalAttempts = 0;
    
    while(min <= max){
        guess = Math.floor((max + min)/2);
        totalAttempts++;
        
        if (array[guess] === targetValue){
            console.log('Total de intentos: ' + totalAttempts);
            return guess;
        }
        else if (array[guess] < targetValue){
            min = guess + 1;
        }
        else{
            max = guess -1;
        }
    }
    return -1;
};

var primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97];

// Comprobación de resultados
var result = doSearch(primes, 73);
console.log("Found prime at index " + result);

console.log(doSearch(primes, 73), 20);

var result1 = doSearch(primes, 11);
console.log("Found prime at index " + result1);

console.log(doSearch(primes, 11),4);

var result2 = doSearch(primes, 97);
console.log("Found prime at index " + result2);

console.log(doSearch(primes, 97),24);
